feat(api/repo): Allow to config repository implementation

laipz8200 opened the pull request:
> [!IMPORTANT]
>
> 1. Make sure you have read our [contribution guidelines](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md)
> 2. Ensure there is an associated issue and you have been assigned to it
> 3. Use the correct syntax to link this PR: `Fixes #<issue number>`.

## Summary

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->
related #20147 

## Screenshots

| Before | After |
|--------|-------|
| ...    | ...   |

## Checklist

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

laipz8200 committed:
feat(api/repo): Allow to config repository implementation

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
feat: Create a DifyAPIRepositoryFactory to handle workflow node execution operations out of core.

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
feat: Create a APIWorkflowRunRepository

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
test(workflow_service): Fix test

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
chore: add new configs to env example

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 commented:
**Update:** Nearly finished, with only some code related to `WorkflowRun` remaining.

laipz8200 committed:
test(test_workflow_draft_variable_service): Fix tests

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 commented:
Hi @QuantumGhost. Do you have some time to take a look at this?

laipz8200 committed:
feat: Integrates execution repository inheritance

Signed-off-by: -LAN- <laipz8200@outlook.com>

QuantumGhost commented on api/tasks/remove_app_and_related_data_task.py:
```diff
@@ -175,30 +177,32 @@ def del_workflow(workflow_id: str):
-        del_workflow_run,
-        "workflow run",
+    """Delete all workflow runs for an app using the service repository."""
+    session_maker = sessionmaker(bind=db.engine, expire_on_commit=False)
```
The `expire_on_commit=False` argument appears to be unnecessary here since no ORM models have been loaded into the session.

QuantumGhost reviewed (dismissed):
LGTM

QuantumGhost commented:
By the way, now that PostgreSQL is available in the CI pipeline, you may write integration tests for the repository implementations to ensure they behave as expected.

laipz8200 committed:
refactor(token_buffer_memory): Convert ValueError to AssertionError

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
refactor(workflow_response_converter): Remove unused workflow_execution_repository

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 commented on api/core/repositories/factory.py:
```diff
@@ -0,0 +1,215 @@
+            raise RepositoryImportError(f"Cannot import repository class '{class_path}': {e}") from e
+
+    @staticmethod
+    def _validate_repository_interface(repository_class: Any, expected_interface: Any) -> None:
```
> (I have already made this change in my fork and can push it to the source branch if we agree on the change.)

SGTM.

laipz8200 committed:
refactor(factory): Move import statement to the top

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 commented on api/core/repositories/factory.py:
```diff
@@ -0,0 +1,215 @@
+        except RepositoryImportError:
+            # Re-raise our custom errors as-is
+            raise
+        except Exception as e:
```
It's not a concern to identify the original error since we utilize `raise from` in the conversation.

laipz8200 committed:
docs(api_workflow_node_execution_repository): Add notice about datat isolation

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
refactor(api_workflow_node_execution_repository): Refactors workflow run query for clarity

Signed-off-by: -LAN- <laipz8200@outlook.com>

QuantumGhost committed:
refactor(api): Simplify the constructor for WorkflowDraftVariableService

since we can utilize `Session.get_bind` to retrieve the correspond `Engine` for a session, there's no need to
manually inject the `sessionmaker` for `WorkflowDraftVariableService`.

QuantumGhost committed:
chore(api): improve type hints for `DifyCoreRepositoryFactory`

QuantumGhost committed:
test(api): fix tests broken by refactor

laipz8200 committed:
refactor(remove_app_and_related_data_task): Remove 'expire_on_commit' to prevent session issues

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
feat(api/repositories/sqlalchemy_api_workflow_run_repository.py): Improves pagination logic for workflow runs

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
docs(api/repositories/api_workflow_node_execution_repository.py): Add note on tenant isolation responsibility

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
docs(api/repositories/sqlalchemy_api_workflow_node_execution_repository.py): Add tenant isolation warning in method docstring

Signed-off-by: -LAN- <laipz8200@outlook.com>

QuantumGhost reviewed (approved):
LGTM
